### PR TITLE
Fix Download for IE11

### PR DESCRIPTION
**Pull request by iyovcheva**
This fixes Invoke/Download errors on IE11

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
[#537](https://github.com/openfaas/faas/issues/537)

## How Has This Been Tested?

* IE11
* Safari 11.0.3
* Firefox 58.0.2
* Chrome  64.0.3282.186 
* Microsoft Edge 40.15063.0.0

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: gateway/assets/script/bootstrap.js [hunk 1]
```diff
@@ -320,8 +320,9 @@ app.controller("home", ['$scope', '$log', '$http', '$location', '$interval', '$f
 ]);
 
 function uuidv4() {
+    var cryptoInstance = window.msCrypto ? window.msCrypto : crypto;
```
Comment (alexellis) [inline gateway/assets/script/bootstrap.js]: This was my input.. @kenfdev does it look OK to you?

File: gateway/assets/script/bootstrap.js [hunk 2]
```diff
@@ -96,11 +96,18 @@ app.controller("home", ['$scope', '$log', '$http', '$location', '$interval', '$f
                     linkElement.setAttribute('href', url);
                     linkElement.setAttribute("download", filename);
          
-                    var clickEvent = new MouseEvent("click", {
-                        "view": window,
-                        "bubbles": true,
-                        "cancelable": false
-                    });
+                    var clickEvent;
```
Comment (alexellis) [inline gateway/assets/script/bootstrap.js]: @iyovcheva please could you explain the new change here?

What do the magic numbers / flags mean or do when combined? i.e.

`"click",true,false,window,0,0,0,0,0,false,false,false,false,0,null`

File: gateway/assets/script/bootstrap.js (hunk 1 above)
Comment (iyovcheva) [reply gateway/assets/script/bootstrap.js]: This should also work

```
var cryptoInstance = window.msCrypto || crypto;
```

File: gateway/assets/script/bootstrap.js (hunk 2 above)
Comment (iyovcheva) [reply gateway/assets/script/bootstrap.js]: @alexellis 

```
object.initMouseEvent (eventName, bubbles, cancelable, view, detail, screenX, screenY, clientX, clientY, ctrlKey, altKey, shiftKey, metaKey, button, relatedTarget);
```

will add a comment in code

File: gateway/assets/script/bootstrap.js (hunk 1 above)
Comment (kenfdev) [reply gateway/assets/script/bootstrap.js]: Both would work but I'd go with 

```
var cryptoObj = window.crypto || window.msCrypto; // for IE 11
```

as the MDN docs says.

https://developer.mozilla.org/en-US/docs/Web/API/Window/crypto

File: gateway/assets/script/bootstrap.js (hunk 2 above)
Comment (iyovcheva) [reply gateway/assets/script/bootstrap.js]: Addressed. Modification is on another line.

File: gateway/assets/script/bootstrap.js (hunk 1 above)
Comment (alexellis) [reply gateway/assets/script/bootstrap.js]: I tried that approach and it failed to work in IE11 giving an exception, this is how we ended up with the above.

Comment (iyovcheva) [reply gateway/assets/script/bootstrap.js]: @alexellis I did a test with the latest changes on all 4 listed browsers&versions. What was the exception?

**Comment by kenfdev** [comment]
@iyovcheva You should use `navigator.msSaveOrOpenBlob` to hack the IE way. The code below should work. (Tested in IE11 and Chrome)

```
var tryDownload = function(data, filename) {
    var caught;

    try {
        var blob = new Blob([data], { type: "binary/octet-stream" });

        if (window.navigator.msSaveBlob) { // // IE hack; see http://msdn.microsoft.com/en-us/library/ie/hh779016.aspx
            window.navigator.msSaveOrOpenBlob(blob, filename);
        }
        else {
            var linkElement = window.document.createElement("a");
            linkElement.href = window.URL.createObjectURL(blob);
            linkElement.download = filename;
            document.body.appendChild(linkElement);
            linkElement.click();
            document.body.removeChild(linkElement);
        }

    } catch (ex) {
        caught = ex;
    }
    return caught;
}
```

FYI
https://stackoverflow.com/a/46233123/6822663

**Comment by alexellis** [comment]
Please could you push up a new gateway image for me to test again on IE11 with your specific fix?

**Comment by kenfdev** [comment]
@alexellis pushed it to `kenfdev/gateway:latest-dev` . This image contains the fix and I PRed it here.
https://github.com/iyovcheva/faas/pull/1

**Comment by iyovcheva** [comment]
Thanks @kenfdev 
I can also test within a day and confirm the fix.

**Comment by iyovcheva** [comment]
Derek remove label: no-dco

**Comment by alexellis** [comment]
Hang on.. why did you ask Derek to remove the label?

**Comment by iyovcheva** [comment]
@kenfdev @alexellis Tested the changes, works well

**Comment by alexellis** [comment]
OK since you've both given sign-off on testing and the changes I'll now merge.

**Comment by alexellis** [comment]
Merged - thank you for the help with this.

**Comment by kenfdev** [comment]
Thank you for testing and reviewing!

**Pull request merged**